Ls4

naokinaokiboo opened the pull request:
lsコマンドの -l オプション対応をpushしました。
ご確認よろしくお願い致します。

naokinaokiboo committed:
mac拡張属性を取得するための拡張ライブラリを追加

naokinaokiboo committed:
lsコマンド -l オプションに対応

naokinaokiboo committed:
コメントアウト行を削除

summertree128 commented on 04.ls/ls.rb:
```diff
@@ -16,13 +42,21 @@ def main
+  end
+end
+
+def dispatch_disp_format(files, long_format, directory = nil)
```
あくまで主観的な意見ですが…プログラム中でdispatchという単語を見ると、言語処理系のやるメソッドディスパッチかOSのタスクディスパッチのどちらかをを想起しがちというか、アプリケーションのコードではあまり使わない単語な気がしますね。（経験上そうでもなかったりしますか？）
まあ使うにしても、フラグ一つを見てif分岐するメソッドの名前に使うにはちょっと仰々しすぎるかなーと感じました。絶対にメソッド名を変えてほしいというほどではないですが、こういう意見もあるということでご参考まで。

naokinaokiboo replied:
> 言語処理系のやるメソッドディスパッチかOSのタスクディスパッチのどちらかをを想起しがちというか、アプリケーションのコードではあまり使わない単語な気がしますね。（経験上そうでもなかったりしますか？）

いえ、仰るとおりだと思います。普通はそういった処理を思い浮かべてしまいますよね。😅
コードを読む側からすると、ややこしさしか無いので修正します。🙇‍♂️

summertree128 commented on 04.ls/ls.rb:
```diff
@@ -59,4 +93,113 @@ def sort_with_reverse_option(array, reverse)
 end
 
+def display_with_long_format(files, directory)
+  file_stat_hash = generate_file_stat_hash(files, directory)
```
変数名に`hash`のような型情報を持たせることはあまりやらない…というかなるべく避けた方が良いとされています。
今回のようなケースなら単に複数形にするだけで十分かと思います。
（型というかデータ構造が重要な場合は変数名にその情報を含めた方が良い場合もありますが、今回は該当しないと考えます）

summertree128 commented on 04.ls/ls.rb:
```diff
@@ -59,4 +93,113 @@ def sort_with_reverse_option(array, reverse)
+end
+
+def generate_file_stat_hash(files, directory)
+  file_stat_hash = {}
+  files.each do |file|
+    file_stat = File.lstat(directory.nil? ? file : [directory, file].join('/'))
+    file_stat_hash[file] = file_stat
+  end
+  file_stat_hash
```
このままでも問題ないですが、Rubyっぽい書き方としてこういう場合は`each_with_object`を使うと103行目や108行目のような記述が不要になりスッキリします。こんな感じです。

```ruby
  files.each_with_object({}) do |file, hash|
    file_stat = File.lstat(directory.nil? ? file : [directory, file].join('/'))
    hash[file] = file_stat
  end
```

何だったら１行でも書けますね。（２行で書く方が好みではありますが）

```ruby
  files.each_with_object({}){|file, hash| hash[file] = File.lstat(directory.nil? ? file : [directory, file].join('/'))}
```

ご参考まで〜

naokinaokiboo replied:
レビューで指摘して頂かないと、ずっと前の書き方のままだと思うので、とても勉強になります！😆
（ちょうど特にこの108行目のような行、前から「書きたくないな〜」と思っていたので😅）

summertree128 commented on 04.ls/ls.rb:
```diff
@@ -59,4 +93,113 @@ def sort_with_reverse_option(array, reverse)
+end
+
+def output_total_block_size(file_stat_hash)
+  total_blocks = 0
+  file_stat_hash.each do |_file, file_stat|
+    total_blocks += file_stat.blocks
+  end
```
このままでも問題ないですが、Rubyっぽい書き方としてこういう場合は`inject`を使うと少し記述が減ってスッキリします。こんな感じです。

```ruby
  total_blocks = file_stat_hash.values.inject(0){|result, file_stat| result += file_stat.blocks }
```

ご参考まで〜

naokinaokiboo replied:
こちらも勉強になります！😆

summertree128 commented on 04.ls/ls.rb:
```diff
@@ -59,4 +93,113 @@ def sort_with_reverse_option(array, reverse)
+    detail_string += get_perission_str(file_stat.mode)
+    file_path = directory.nil? ? file : [directory, file].join('/')
+    mac_xattr = MacXattr.new
+    detail_string += mac_xattr.get_macxattr(file_path) + delimiter
+    detail_string += file_stat.nlink.to_s.rjust(nlink_length) + delimiter
+    detail_string += Etc.getpwuid(file_stat.uid).name.rjust(owner_name_length) + delimiter * 2
+    detail_string += Etc.getgrgid(file_stat.gid).name.rjust(group_name_length) + delimiter * 2
+    detail_string += file_stat.size.to_s.rjust(file_size_length) + delimiter
+    detail_string += get_modified_time_string(file_stat) + delimiter
+    detail_string += file_stat.symlink? ? "#{file} -> #{File.readlink(file_path)}" : file
```
これまたそのままでも問題ないと言えばないのですが、文字列連結の方法としては`+=`より`<<`の方が早いらしいので、僕はなるべくこちらを使っています。（非破壊的にする必要がある場合を除く）

参考：https://simanman.hatenablog.com/entry/2014/06/23/002908

summertree128 reviewed (commented):
レビューしました！

拡張属性、こうやって実装すれば対応できるんですね…。やったことありませんでした。
C言語は詳しくないので拡張ライブラリの方はちょっとレビューできませんが、Rubyの方にはいくつかコメントをつけました。
とはいえ、特にコメントなしでOKにしてもいいくらいのレベルではあるのですが…。
一応、参考になればな〜という感じのことを書いています。変数名/メソッド名のところは一応再考してみてもらいたいです。よろしくお願いします。

naokinaokiboo committed:
レビュー指摘事項に対応

・所有者名、グループ名を左寄せに修正
・関数名を修正(dispatch_disp_format -> call_disp_functions)
・変数名を修正(file_stat_hash -> file_stats)
・文字列連結の方法を修正(+= -> <<)
・each_with_object, injectを用いた書き方に修正

summertree128 commented on 04.ls/ls.rb:
```diff
@@ -183,11 +177,9 @@ def get_permission_attr_str(permission_bits)
-  permission_masks.each do |mask, attribute|
-    permission_str += mask & permission_bits != 0 ? attribute : '-'
+  permission_masks.each_with_object(+'') do |(mask, attr), permission_str|
+    permission_str << (mask & permission_bits != 0 ? attr : '-')
```
👍

summertree128 reviewed (approved):
レビューOKです！
